java/src/main/java/ai/rapids/cudf/DType.java

```diff
@@ -17,76 +17,202 @@
+      INT32,
+      INT64,
+      UINT8,
+      UINT16,
+      UINT32,
+      UINT64,
+      FLOAT32,
+      FLOAT64,
+      BOOL8,
+      TIMESTAMP_DAYS,
+      TIMESTAMP_SECONDS,
+      TIMESTAMP_MILLISECONDS,
+      TIMESTAMP_MICROSECONDS,
+      TIMESTAMP_NANOSECONDS,
+      DURATION_DAYS,
+      DURATION_SECONDS,
+      DURATION_MILLISECONDS,
+      DURATION_MICROSECONDS,
+      DURATION_NANOSECONDS,
+      null, // DICTIONARY32
+      STRING,
+      LIST,
+      null, // DECIMAL32
+      null, // DECIMAL64
+      STRUCT
+  };
+
   /**
-   * ms since the UNIX epoch
+   * This only works for fixed width types. Variable width types like strings the value is
+   * undefined and should be ignored.
+   *
+   * @return size of type in bytes.
    */
-  TIMESTAMP_MILLISECONDS(8, 14, "timestamp[ms]"),
+  public int getSizeInBytes() { return typeId.sizeInBytes; }
+
   /**
-   * microseconds since the UNIX epoch
+   * Returns
+   * @return
    */
-  TIMESTAMP_MICROSECONDS(8, 15, "timestamp[us]"),
+  public int getNativeId() { return typeId.nativeId; }
+
   /**
-   * ns since the UNIX epoch
+   * Returns scale for Decimal Type
+   * @return scale
    */
-  TIMESTAMP_NANOSECONDS(8, 16, "timestamp[ns]"),
-
-  //We currently don't have mappings for duration type to I/O files, and these
-  //simpleNames might change in future when we do
-  DURATION_DAYS(4, 17, "int32"),
-  DURATION_SECONDS(8, 18, "int64"),
-  DURATION_MILLISECONDS(8, 19, "int64"),
-  DURATION_MICROSECONDS(8, 20, "int64"),
-  DURATION_NANOSECONDS(8, 21, "int64"),
-  //DICTIONARY32(4, 22, "NO IDEA"),
-
-  STRING(0, 23, "str"),
-  LIST(0, 24, "list"),
-  STRUCT(0, 27, "struct");
-
-  private static final DType[] D_TYPES = DType.values();
-  final int sizeInBytes;
-  final int nativeId;
-  final String simpleName;
-
-  DType(int sizeInBytes, int nativeId, String simpleName) {
-    this.sizeInBytes = sizeInBytes;
-    this.nativeId = nativeId;
-    this.simpleName = simpleName;
+  public int getScale() { return scale; }
+
+  public String getSimpleName() { return typeId.simpleName; }
+
+  // TODO: override hashCode
+  @Override
+  public boolean equals(Object o) {
+    DType other = (DType) o;
+    return this.typeId == other.typeId && this.scale == other.scale;
   }
 
-  public boolean isTimestamp() {
-    return TIMESTAMPS.contains(this);
+  public static DType fromNative(int nativeId, int scale) {
+    if (nativeId >=0 && nativeId < staticDTypes.length) {
+      if (staticDTypes[nativeId] != null) {
+        return staticDTypes[nativeId];
+      }
+      if (nativeId == DTypeEnum.DECIMAL32.nativeId) {
+        return new DType(DTypeEnum.DECIMAL32, scale);
+      }
+      if (nativeId == DTypeEnum.DECIMAL64.nativeId) {
+        return new DType(DTypeEnum.DECIMAL64, scale);
+      }
+    }
+    throw new IllegalArgumentException("Could not translate " + nativeId + " into a DType");
+  }
+
+  public static DType fromNative(int nativeId) {
+    return fromNative(nativeId, Integer.MIN_VALUE);
```

Comment:
IMO we should not be implicitly setting the scale of decimal types.  Trying to convert from a native ID to a decimal type without specifying a scale should be an error.